packages/fxa-auth-server/lib/routes/recovery-codes.js

```diff
@@ -65,6 +65,99 @@ module.exports = (log, db, config, customs, mailer) => {
         return { recoveryCodes };
       },
     },
+    {
+      method: 'GET',
+      path: '/newRecoveryCodes',
+      options: {
+        auth: {
+          strategy: 'sessionToken',
+        },
+        response: {
+          schema: {
+            recoveryCodes: isA.array().items(isA.string())
+          }
+        }
+      },
+      async handler(request) {
+        log.begin('newRecoveryCodes', request);
+
+        const { authenticatorAssuranceLevel, uid } = request.auth.credentials;
+
+        // Since TOTP and recovery codes go hand in hand, you should only be
+        // able to replace recovery codes in a TOTP verified session.
+        if (!authenticatorAssuranceLevel || authenticatorAssuranceLevel <= 1) {
+          throw errors.unverifiedSession();
+        }
+
+        const recoveryCodes = await db.createRecoveryCodes(
+          uid,
+          RECOVERY_CODE_COUNT
+        );
+
+        await request.emitMetricsEvent('recoveryCode.created', { uid });
+
+        return { recoveryCodes };
+      }
+    },
+    {
+      method: 'POST',
+      path: '/updateRecoveryCodes',
+
+      options: {
+        auth: {
+          strategy: 'sessionToken',
+        },
+        validate: {
+          payload: {
+            recoveryCodes: isA.array().items(isA.string()),
+          },
+        },
+        response: {
+          schema: {
+            success: isA.boolean(),
+          }
+        }
```

Comment:
I thought about this a bit more...

I think we're actually ok with accepting values from the client as long as we can validate them since we require a verified session, the codes are random and not connected to any other data, and we send a confirmation email. A round trip with a signed JWT is overkill for this.

Our validation should be more thorough than what's here so far. We should ensure they are the correct length and valid base32, no duplicates.

And since we're allowing client values we don't really need a server endpoint to generate them. We can use https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues on the client-side